File: src/main/java/com/example/chatrooms/room/RoomService.java
package com.example.chatrooms.room;

import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class RoomService {
    private final RoomRepository roomRepository;
    private final RoomSessionService roomSessionService;

    public RoomService(RoomRepository roomRepository, RoomSessionService roomSessionService){
        this.roomRepository = roomRepository;
        this.roomSessionService = roomSessionService;
    }

    public RoomSummary createRoom(String rawName){
        String trimmedName = rawName.trim();

        if(roomRepository.existsByNameIgnoreCase(trimmedName)){
            throw new RoomNameTakenException();
        }
        Room saved = roomRepository.save(new Room(trimmedName));
        return this.toSummary(saved);
    }

    public List<RoomSummary> listRooms(){
        return roomRepository.findAll().stream().map(this::toSummary).toList();
    }

    public RoomSummary getRoom(String name){
        return roomRepository.findByNameIgnoreCase(name.trim()).map(this::toSummary).orElseThrow(RoomNotFoundException::new);
    }

    private RoomSummary toSummary(Room room){
        int count  = (int) roomSessionService.countInRoom(room.getName());
        return new RoomSummary(room.getName(), room.getCreatedAt(),count);
    }
}

File: frontend/src/app/room.service.ts
import { Service, Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { RoomSummary } from './room';

@Service()
export class RoomService {
  private http = inject(HttpClient);

  listRooms(): Observable<RoomSummary[]>{
    return this.http.get<RoomSummary[]>('/rooms');
  }

  createRoom(name: string):Observable<RoomSummary>{
    return this.http.post<RoomSummary>('/rooms', {name});
  }

  getRoom(name: string) : Observable<RoomSummary>{
    return this.http.get<RoomSummary>('/rooms/'+encodeURIComponent(name));
  }
}
